File: Services/ComissaoService.cs
using DesafioTarget.API.Models;

namespace DesafioTarget.API.Services;

    public class ComissaoService
{
    public List<VendedorComissao> CalcularComissoes(List<Venda> vendas)
    {
        var vendasPorVendedor = vendas.GroupBy(v => v.Vendedor);

        var resultado = new List<VendedorComissao>();

        foreach (var grupo in vendasPorVendedor)
        {
            var VendedorComissao = new VendedorComissao
            {
                Vendedor = grupo.Key,
                VendasComComissao = new List<VendaComissao>()
            };

            foreach (var venda in grupo)
            {
                var comissao = CalcularComissaoIndividual(venda.Valor);
                var percentual = ObterPercentualComissao(venda.Valor);

                VendedorComissao.VendasComComissao.Add(new VendaComissao
                {
                    ValorVenda = venda.Valor,
                    Comissao = comissao,
                    PercentualComissao = percentual
                });
            };

            resultado.Add(VendedorComissao);
        }

        return resultado;
    }

    private decimal CalcularComissaoIndividual(decimal valorVenda)
    {
        decimal percentual = ObterPercentualComissao(valorVenda);
        return valorVenda * percentual / 100;
    }

    private decimal ObterPercentualComissao(decimal valorVenda)
    {
        if (valorVenda < 100.00m)
        {
            return 0m;
        }

        else if (valorVenda < 500.00m)
        {
            return 1m;
        }
        else
        {
            return 5m;
        }
    }
}

File: src/app/services/comissao.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Venda {
  vendedor: string;
  valor: number;
}

export interface VendaComComissao {
  valorVenda: number;
  comissao: number;
  percentualComissao: number;
}

export interface VendedorComissao {
  vendedor: string;
  vendasComComissao: VendaComComissao[];
}

export interface VendasData {
  vendas: Venda[];
}

@Injectable({
  providedIn: 'root'
})
export class ComissaoService {
  private apiUrl = 'http://localhost:5297/api/comissoes';

  constructor(private  http: HttpClient) { }

  calcularComissoes(vendas: Venda[]): Observable<VendedorComissao[]> {
    const body: VendasData = { vendas };
    return this.http.post<VendedorComissao[]>(`${this.apiUrl}/calcular`, body);
  }
}
